### PR TITLE
Add support for active_kernels

### PR DESCRIPTION
**Pull request by michel2323**
Related to https://github.com/JuliaGPU/AMDGPU.jl/issues/139

### CODE REVIEW

File: src/AMDGPU.jl [hunk 1]
```diff
@@ -153,6 +154,7 @@ function __init__()
                 Please run Pkg.build("AMDGPU") and reload AMDGPU.
                 """
             end
+            active_kernels[get_default_queue()] = Vector{AMDGPU.RuntimeEvent{AMDGPU.HSAStatusSignal}}()
```
Comment (jpsamaroo) [inline src/AMDGPU.jl:157]: This should move to the `HSAQueue` constructor, which will be invoked by `get_default_queue()` on its first call for a given agent.

File: src/execution.jl [hunk 2]
```diff
@@ -193,6 +193,7 @@ macro roc(ex...)
                     foreach($wait!, ($(var_exprs...),))
                     if $launch
                         local $signal = $create_event($kernel.mod.exe)
+                        push!(AMDGPU.active_kernels[AMDGPU.get_default_queue()], $signal)
```
Comment (jpsamaroo) [inline src/execution.jl:196]: This won't work if the queue selected by the user (and passed in `call_kwargs`) isn't the default queue.

File: src/runtime.jl [hunk 3]
```diff
@@ -72,6 +72,7 @@ function Base.wait(event::RuntimeEvent{HSAStatusSignal}; check_exceptions=true,
                 deleteat!(mod.metadata, idx)
             end
         end
+        deleteat!(active_kernels[get_default_queue()], findall(x->x==event, active_kernels[get_default_queue()]))
```
Comment (jpsamaroo) [inline src/runtime.jl:75]: We should probably get the queue from the event. I'm not sure if our `HSAStatusSignal` contains the queue it was launched on, so we'll possibly have to add that as a field.

**Review by jpsamaroo — COMMENTED** [review]
:heart:

**Comment by jpsamaroo** [comment]
I'm trying to push changes to your branch, but I'm getting a permissions error. Did you enable the option to let maintainers push to your branch?

**Comment by michel2323** [comment]
For some reason I don't see the "Allow edits from maintainers" option. I just gave you admin rights.

**Comment by jpsamaroo** [comment]
Thanks, pushed!

**Comment by jpsamaroo** [comment]
I hope you don't mind, I'm going to use this PR to rip out some old functionality that just complicates the code base for no tangible benefit.

**Comment by michel2323** [comment]
Of course not!

**Comment by jpsamaroo** [comment]
I've modified the API of `HSAStatusSignal` so that each signal always generates a background waiter which waits on the real HSA signal, and notifies tasks that called `wait` on the signal once that's finished. All foreground waiting tasks will also throw a `KernelException` if the kernel threw an exception.

This background task will also remove the kernel from the set of active kernels. @michel2323 I added a runtime lock, which will need to be taken to manipulate `_active_kernels`, and exposed the function `active_kernels()`, which is now the user-facing API.

**Comment by jpsamaroo** [comment]
Ok, just need to pass the new test, and add docs. @michel2323 the API has been changed to accept a queue argument, and will default to the default queue.

**Comment by jpsamaroo** [comment]
Thanks a ton @michel2323 !

**Pull request merged**